### ISSUE TITLE
ElasticDL Optimizer only Supports Sparse Update for Embedding Layer

### ISSUE THREAD

**Issue by LiMinghao1994** [sql-machine-learning/elasticdl#1083]
After PR #1079, we only support sparse update for ElasticDL's embedding layer. Here sparse means updating only the parameters and slots appear in this batch. This is called *sparse update* or *lazy update* in PyTorch and in `tf.contrib` in TenforFlow 1.x.

TensorFlow adopts dense update for some optimizers, e.g. Adam optimizer. Because the update rules of optimization algorithms are comprised of slots and gradients. When slots do not equal to zero, model parameters can change even if gradients equal to zero.

Although TensorFlow adopts dense update for its Adam optimizer, sparse version is also widely used. There is a sparse version called `tf.contrib.LazyAdamOptimizer` in TensorFlow 1.x. TensorFlow 2.x remove the `contrib` and [tensorflow addons](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md) implements `lazy_adam` for TensorFlow 2.x. PyTorch implements both `Adam` and `SparseAdam` as its official optimizers.

For the performance reason, we choose to implement only the sparse version of optimizers. If we implement dense version optimizers, we should lookup and update the whole embedding table every time we need to update model.

**Comment by mhaoli** [comment]
TensorFlow adopts dense updating rule for optimizer `Adam`, `Nadam` and `RMSprop`. In ElasticDL, currently we only implement sparse updating rule if model has ElasticDL embedding layer. We don't plan to implement dense updating rule in the short term.
